src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java

```diff
@@ -0,0 +1,343 @@
+			list.add(te);
+			content.put(pos, list);
+			return true;
+		}
+		else
+			return list.add(te);
+	}
+	
+	@Override
+	public void add(int index, TileEntity te) {
+		add(te);
+	}
+	
+	@Override
+	public boolean addAll(int index, Collection<? extends TileEntity> tileEntities) {
+		return addAll(tileEntities);
+	}
+	
+	@Override
+	public boolean removeAll(Collection<?> tileEntities)
+	{
+		boolean removedAll = true;
+		for (Iterator<?> iterator = tileEntities.iterator(); iterator.hasNext();) {
+			if(!remove(iterator.next()))
+				removedAll = false;
+		}
+		return removedAll;
+	}
+	
+	@Override
+	public boolean remove(Object te)
+	{
+		if(!(te instanceof TileEntity))
+			return false;
+		ChunkPos pos = getChunkPos((TileEntity) te);
+		List<TileEntity> values = getChunkTEList(pos);
+		if(values != null)
+			if(values.remove(te))
+			{
+				if(values.isEmpty())
+					removeChunk(pos);
+				return true;
+			}
+		return false;
+	}
+	
+	private boolean removeChunk(ChunkPos key)
+	{
+		return content.remove(key) != null;
+	}
+	
+	public boolean removeChunk(World world, ChunkPos pos, boolean notify)
+	{
+		if(notify)
+		{
+			for (TileEntity te : getChunkTEList(pos))
+	        {
+				te.onChunkUnload();
+	        }
+		}
+		
+		return removeChunk(pos);
+	}
+	
+	public int totalCount()
+	{
+		int size = 0;
+		for (List<TileEntity> values : content.values()) {
+			size += values.size();
+		}
+		return size;
+	}
+	
+	public int chunkCount()
+	{
+		return content.size();
+	}
+	
+	@Override
+	public void clear()
+	{
+		content.clear();
+	}
+	
+	@Override
+	public String toString()
+	{
+		return content.toString();
+	}
+	
+	@Override
+	public boolean isEmpty() {
+		return totalCount() > 0;
+	}
+	
+	@Override
+	public Iterator<TileEntity> iterator()
+	{
+		return new Iterator<TileEntity>() {
+			
+			int index = 0;
+			
+			Iterator<List<TileEntity>> iterator = values().iterator();
+			
+			List<TileEntity> currentList;
+			
+			@Override
+			public boolean hasNext() {
+				while(currentList == null || currentList.size() <= index)
+				{
+					if(iterator.hasNext())
+					{
+						currentList = iterator.next();
+						index = 0;
+					}
+					else
+						return false;
+				}
+				
+				return true;
+			}
+
+			@Override
+			public TileEntity next() {
+				TileEntity value = currentList.get(index);
+				index++;
+				return value;
+			}
+			
+			@Override
+			public void remove() {
+				currentList.remove(index-1);
+			}
+		};
+	}
+
+	@Override
+	public boolean retainAll(Collection<?> paramCollection) {
+		boolean changed = false;
+		for (List<TileEntity> values : content.values()) {
+			if(values.retainAll(paramCollection))
+				changed = true;
+		}
+		return changed;
+	}
+
+	@Override
+	public int size() {
+		return totalCount();
+	}
+
+	@Override
+	public Object[] toArray() {
+		Object[] array = new Object[totalCount()];
+		int i = 0;
+		for (List<TileEntity> values : content.values()) {
+			for (int j = 0; j < values.size(); j++) {
```

Comment:
that would work, but why create the confusion? I mean, is there any benifit of performance involved? To my knoweledge it will create an Interator which will slow things down beside creating confusion.